mlir/lib/Quantum/Transforms/ChainedSelfInversePatterns.cpp

```diff
@@ -56,9 +56,15 @@ struct ChainedNamedHermitianOpRewritePattern : public mlir::OpRewritePattern<Cus
 
         // Replace uses
         ValueRange InQubits = op.getInQubits();
-        auto ParentOp = dyn_cast_or_null<CustomOp>(InQubits[0].getDefiningOp());
-        ValueRange simplifiedVal = ParentOp.getInQubits();
-        rewriter.replaceOp(op, simplifiedVal);
+        auto parentOp = dyn_cast_or_null<CustomOp>(InQubits[0].getDefiningOp());
+        ValueRange originalNonCtrlQubits = parentOp.getNonCtrlQubitOperands();
+        ValueRange originalCtrlQubits = parentOp.getCtrlQubitOperands();
+        for (const auto &[idx, nonCtrlQubitResult] : llvm::enumerate(op.getNonCtrlQubitResults())) {
+            nonCtrlQubitResult.replaceAllUsesWith(originalNonCtrlQubits[idx]);
+        }
+        for (const auto &[idx, ctrlQubitResult] : llvm::enumerate(op.getCtrlQubitResults())) {
+            ctrlQubitResult.replaceAllUsesWith(originalCtrlQubits[idx]);
+        }
```

Comment:
Can you try this?

```suggestion
        auto parentOp = dyn_cast_or_null<CustomOp>(InQubits[0].getDefiningOp());
        ValueRange qubitOperands = parentOp.getQubitOperands();
        rewriter.replaceOp(op, qubitOperands)'
        return success();
```

Reply:
@dime10 is there a reason for this difference :thinking:

Reply:
`getQubitOperands` is an interface method. Some quantum operations are not unitary (and thus cannot be controlled/adjointed), so there is a base interface for that, while the `QuantumGate` interface would include control qubits as well.

`CustomOp` should obey the `QuantumGate` interface.

Reply:
I would prefer keeping the control and non control qubits explicit, if that's ok by you.

Reply:
Why?

Reply:
I agree that Erick's version is much more succinct, not only avoiding the ctrl/non-ctrl split, but also the for loops.

Reply:
My concern is without this being explicit, the tablegen methods need to keep in mind that both `getQubitOperands` and `getQubitResults` should be in the same order of "non ctrl bits, ctrl bits"; otherwise there would be errors in this one-line replacement, leading either to a crash or a logic error when the number of wires happens to not crash.

Reply:
Or, the `getQubitOperands` and the default op results need to be in the same order, but the order of `getQubitOperands` is not mandated, or some other order might be desired by other passes.

Reply:
> My concern is without this being explicit, the tablegen methods need to keep in mind that both getQubitOperands and getQubitResults should be in the same order of "non ctrl bits, ctrl bits"

Absolutely, but that is on the method implementations to ensure. The whole point of this particular interface method is that you don't need to care what is what, you can simply ask for all qubit inputs or outputs.

> Or, the getQubitOperands and the default op results need to be in the same order, but the order of getQubitOperands is not mandated, or some other order might be desired by other passes.

It's true that `rewriter.replaceOp(op, qubitOperands)` makes some additional assumptions, mainly that:
- the op only has qubit results, which I think is implied by the interface description (at least its true for all the ops we have)
- the op result order matches the order of `getQubitResults`. This is technically not stated anywhere, although it is true for all our current ops. While this might be an argument for looping over and replacing each qubit separately, it wouldn't warrant treating ctrl qubits differently than non-ctrl qubits.

Reply:
```
        ValueRange InQubits = op.getInQubits();
        auto parentOp = cast<CustomOp>(InQubits[0].getDefiningOp());
        ValueRange qubitOperands = parentOp.getQubitOperands();
        for (const auto &[idx, qubitResult] : llvm::enumerate(op.getQubitResults())) {
            qubitResult.replaceAllUsesWith(qubitOperands[idx]);
        }
```

This breaks:
```
quantum-opt: /home/paul.wang/catalyst_new/catalyst/mlir/llvm-project/mlir/include/mlir/IR/UseDefLists.h:213: void mlir::IRObjectWithUseList<mlir::OpOperand>::replaceAllUsesWith(ValueT &&) [OperandType = mlir::OpOperand, ValueT = mlir::Value &]: Assertion `(!newValue || this != OperandType::getUseList(newValue)) && "cannot RAUW a value with itself"' failed.
```

I'm looking into why...

Reply:
(Note that the version on the PR, where ctrl/non-ctrl are separated, works fine)

Reply:
hmm, 
```
func.func @test_chained_self_inverse() -> !quantum.bit {
    %0 = quantum.alloc( 1) : !quantum.reg
    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
    %2 = quantum.custom "Hadamard"() %1 : !quantum.bit
    %3 = quantum.custom "Hadamard"() %2 : !quantum.bit
    return %3 : !quantum.bit
}
```

```C
        ValueRange InQubits = op.getInQubits();
        auto parentOp = cast<CustomOp>(InQubits[0].getDefiningOp());
        ValueRange qubitOperands = parentOp.getQubitOperands();
        for (const auto &[idx, qubitResult] : llvm::enumerate(op.getQubitResults())) {
            llvm::errs() << qubitResult << "\n";
            llvm::errs() << qubitOperands[idx] << "\n";
            qubitResult.replaceAllUsesWith(qubitOperands[idx]);
        }
```

```
%out_qubits_0 = quantum.custom "Hadamard"() %out_qubits : !quantum.bit
%out_qubits_0 = quantum.custom "Hadamard"() %out_qubits : !quantum.bit

error: .. cannot RAUW a value with itself
```